lib/api/apiUtils/object/corsResponse.js

```diff
@@ -0,0 +1,127 @@
+* @param {string[]} [rules[].allowedHeaders] - headers allowed in an
+*   OPTIONS request via the Access-Control-Request-Headers header
+* @param {number} [rules[].maxAgeSeconds] - seconds browsers should cache
+*   OPTIONS response
+* @param {string[]} [rules[].exposeHeaders] - headers to expose to external
+*   applications
+* @param {string} origin - origin of CORS request
+* @param {string} method - expected method of CORS request
+* @param {string[]} [headers] - expected headers of CORS request
+* @return {(null|object)} - matching rule if found; null if no match
+*/
+export function findCorsRule(rules, origin, method, headers) {
+    for (let i = 0; i < rules.length; i++) {
+        const rule = rules[i];
+        if (rule.allowedMethods.indexOf(method) === -1) {
+            continue;
+        } else if (!_matchesOneOf(rule.allowedOrigins, origin)) {
+            continue;
+        } else if (headers &&
+            !_headersMatchRule(headers, rule.allowedHeaders)) {
+            continue;
+        }
+        // if it passed all above checks, found matching rule
+        return rule;
+    }
+    // searched all rules and found no match
+    return null;
+}
+
+/** _gatherResHeaders - Collect headers to return in response
+* @param {object} rule - array of rules
+* @param {string} [rule.id] - optional id to identify rule
+* @param {string[]} rule[].allowedMethods - methods allowed for CORS
+* @param {string[]} rule[].allowedOrigins - origins allowed for CORS
+* @param {string[]} [rule[].allowedHeaders] - headers allowed in an
+*   OPTIONS request via the Access-Control-Request-Headers header
+* @param {number} [rule[].maxAgeSeconds] - seconds browsers should cache
+*   OPTIONS response
+* @param {string[]} [rule[].exposeHeaders] - headers to expose to external
+*   applications
+* @param {string} origin - origin of CORS request
+* @param {string} method - expected method of CORS request
+* @param {string[]} [headers] - expected headers of CORS request
+* @return {object} resHeaders - headers to include in response
+*/
+export function gatherCorsResHeaders(rule, origin, method, headers) {
+    const resHeaders = {
+        'access-control-max-age': rule.maxAgeSeconds,
+        'access-control-allow-methods': rule.allowedMethods.join(', '),
+        'content-length': '0',
+        'date': new Date().toUTCString(),
+        'vary':
+        'Origin, Access-Control-Request-Headers, Access-Control-Request-Method',
+    };
+    // send back '*' if any origin allowed; otherwise send back
+    // request Origin value
+    if (rule.allowedOrigins.indexOf('*') > -1) {
+        resHeaders['access-control-allow-origin'] = '*';
+    } else {
+        resHeaders['access-control-allow-origin'] = origin;
+        // can only set allow credentials to true if not returning wildcard
+        // for 'access-control-allow-origin'
+        resHeaders['access-control-allow-credentials'] = true;
```

Comment:
Generally the access-control-allow-credentials header affects requests with a 'Cookie' header. So I added a test with a Cookie header, and the response header/body look the same as it would be for a request without the header. (Even if the access-control-allow-origin is '*' & access-control-allow-credential is omitted)